src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,555 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\String;
+
+use Symfony\Component\String\Exception\ExceptionInterface;
+use Symfony\Component\String\Exception\InvalidArgumentException;
+
+/**
+ * Represents a string.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author Hugo Hamon <hugohamon@neuf.fr>
+ *
+ * @throws ExceptionInterface
+ */
+abstract class AbstractString implements \JsonSerializable
+{
+    protected $string = '';
+    protected $ignoreCase = false;
+    protected $locale = null;
+
+    abstract public function __construct(string $string = '', string $locale = null);
+
+    public function toGrapheme(): GraphemeString
+    {
+        return new GraphemeString($this->string, $this->locale);
+    }
+
+    public function toUtf8(): Utf8String
+    {
+        return new Utf8String($this->string, $this->locale);
+    }
+
+    public function toBinary(): BinaryString
+    {
+        return new BinaryString($this->string, $this->locale);
+    }
+
+    /**
+     * @return static
+     */
+    public function ignoreCase(): self
+    {
+        $str = clone $this;
+        $str->ignoreCase = true;
+
+        return $str;
+    }
+
+    /**
+     * @return static
+     */
+    abstract public function after(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static
+     */
+    abstract public function afterLast(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static
+     */
+    abstract public function append(string ...$suffix): self;
+
+    /**
+     * @return static
+     */
+    abstract public function before(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static
+     */
+    abstract public function beforeLast(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static[]
+     */
+    abstract public function chunk(int $length = 1): array;
+
+    /**
+     * @return static
+     */
+    public function collapseWhitespace(): self
+    {
+        $str = clone $this;
+        $str->string = trim(preg_replace('/(?:\s{2,}+|[^\S ])/', ' ', $str->string));
+
+        return $str;
+    }
+
+    abstract public function equalsTo(string $string): bool;
+
+    abstract public function endsWith(string $suffix): bool;
+
+    /**
+     * @return static
+     */
+    public function ensureStart(string $prefix): self
```

Comment:
@javiereguiluz @apfelbox did I get you right this is what you're suggesting: using start/end instead of left/right?

Reply:
Yes, exactly. This way these methods "read right" even for non-LTR languages. 👍

Reply:
`^` always means the beginning of the string and doesn't care about the directionality, so there is nothing to do, all is fine :)